File: android/app/src/main/java/com/timestampcamerafree/gpsmapcameratimemark/geotagginglocationonphoto/views/widget/CircularCharacterView.kt
package com.timestampcamerafree.gpsmapcameratimemark.geotagginglocationonphoto.views.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.View
import kotlin.math.max

class CircularCharacterView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private var text: String = ""

    private var characterBackgroundColor: Int = Color.BLUE // 默认颜色
    private var characterBackgroundColors: List<Int>? = null // 多颜色支持
    private var textColor: Int = Color.WHITE
    private  var circlePadding:Int = 2
    private  var circleSpacing:Int = 2
    private var textSizePx: Float = spToPx(12f)
    private val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
    }

    //region Public API
     fun setText(text: String) {

        //this.text = text.replace("\\D".toRegex(), "")
         this.text = text.replace("⃣\uFE0F", "")//去掉原来的那个文字的背景和空格
        requestLayout()
        invalidate()
    }

    fun setCharacterBackgroundColor(color: Int) {
        this.characterBackgroundColor = color
        this.characterBackgroundColors = null // 清除每字符颜色
        invalidate()
    }

    fun setCharacterBackgroundColors(colors: List<Int>) {
        this.characterBackgroundColors = colors
        invalidate()
    }
// 默认 16sp
    private fun spToPx(sp: Float): Float {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP,
            sp,
            resources.displayMetrics
        )
    }
    fun setTextColor(color: Int) {
        this.textColor = color
        invalidate()
    }
    //endregion

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val desiredWidth = calculateDesiredWidth()
        val desiredHeight = calculateDesiredHeight()
        //Log.d("calculateDesiredWidth","calculateDesiredWidth ${desiredWidth} ")
        setMeasuredDimension(
            resolveSize(desiredWidth, widthMeasureSpec),
            resolveSize(desiredHeight, heightMeasureSpec)
        )
    }

//    private fun calculateDesiredHeight(): Int {
//        val padding = 6 * resources.displayMetrics.density
//        textPaint.textSize = 48f
//        val fontMetrics = textPaint.fontMetrics
//        val textHeight = fontMetrics.bottom - fontMetrics.top
//        val radius = (textHeight / 2 + padding)
//        return (radius * 2).toInt()
//    }
    private fun calculateRadius(): Float {
        val padding = circlePadding * resources.displayMetrics.density
        val fontMetrics = textPaint.fontMetrics
        val textHeight = fontMetrics.bottom - fontMetrics.top
        return (max(textSizePx, textHeight) / 2 + padding)
    }
    private fun calculateDesiredWidth(): Int {
        if (text.isEmpty()) return 0

        val radius = calculateRadius()
        val spacing = circleSpacing
        val diameter = radius * 2
        val characterSpace = diameter + spacing

        return (text.length * characterSpace).toInt()
    }
    private fun calculateDesiredHeight(): Int {
        val radius = calculateRadius()
        return (radius * 2).toInt()
    }

    fun setTextSize(sp: Float) {
        this.textSizePx = spToPx(sp)
        textPaint.textSize = textSizePx
        requestLayout()
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (text.isEmpty()) return

        textPaint.color = textColor
        textPaint.textSize = textSizePx
        val radius = calculateRadius()
        val spacing = circleSpacing
        val diameter = radius * 2
        val characterSpace = diameter + spacing
        val startX = (width - (text.length * characterSpace) + spacing) / 2 + radius
        val centerY = height / 2f
        text.forEachIndexed { index, char ->
            val centerX = startX + (index * characterSpace)
            // 设置背景颜色：每字符颜色优先，没有则用默认色
            backgroundPaint.color = characterBackgroundColors?.getOrNull(index) ?: characterBackgroundColor

            canvas.drawCircle(centerX, centerY, radius, backgroundPaint)

            val textY = centerY - ((textPaint.descent() + textPaint.ascent()) / 2)
            canvas.drawText(char.toString(), centerX, textY, textPaint)
        }
    }
}

File: ios/iOSTimeGPS/UI/Watermark/Views/BaseID16/CircularCharacterView.swift
//
//  CircularCharacterView.swift
//  iOSTimeGPS
//
//  Created by waynelu on 2025/6/24.
//

import Foundation
import UIKit

class CircularCharacterView: UIView {

    private var text: String = ""
    private var characterBackgroundColor: UIColor = .blue
    private var characterBackgroundColors: [UIColor]? = nil
    private var textColor: UIColor = .white
    private var textSize: CGFloat = 12.0
    private let circlePadding: CGFloat = 1.0
    private let circleSpacing: CGFloat = 2.0

    private let textAttributes: [NSAttributedString.Key: Any] = [:]

    // MARK: - Public API

    func setText(_ text: String) {
        self.text = text.replacingOccurrences(of: "⃣️", with: "").replacingOccurrences(of:" ", with: "")
        setNeedsDisplay()
        invalidateIntrinsicContentSize()
    }

    func setCharacterBackgroundColor(_ color: UIColor) {
        self.characterBackgroundColor = color
        self.characterBackgroundColors = nil
        setNeedsDisplay()
    }

    func setCharacterBackgroundColors(_ colors: [UIColor]) {
        self.characterBackgroundColors = colors
        setNeedsDisplay()
    }

    func setTextColor(_ color: UIColor) {
        self.textColor = color
        setNeedsDisplay()
    }

    func setTextSize(sp: CGFloat) {
        self.textSize = sp
        setNeedsDisplay()
        invalidateIntrinsicContentSize()
    }

    // MARK: - Size Calculation

    private func calculateRadius() -> CGFloat {
        let padding = circlePadding * UIScreen.main.scale
        let font = UIFont.systemFont(ofSize: textSize)
        let textHeight = font.lineHeight
        return max(textSize, textHeight) / 2 + padding
    }

     func calculateDesiredWidth() -> CGFloat {
        guard !text.isEmpty else { return 0 }
        let radius = calculateRadius()
        let diameter = radius * 2
        let characterSpace = diameter + circleSpacing
        return CGFloat(text.count) * characterSpace
    }

     func calculateDesiredHeight() -> CGFloat {
        return calculateRadius() * 2
    }

    override var intrinsicContentSize: CGSize {
        return CGSize(width: calculateDesiredWidth(), height: calculateDesiredHeight())
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        guard !text.isEmpty else { return }

        let radius = calculateRadius()
        let diameter = radius * 2
        let characterSpace = diameter + circleSpacing
       // let startX = (bounds.width - CGFloat(text.count) * characterSpace + circleSpacing) / 2 + radius
        let startX = radius
        let centerY = bounds.height / 2

        let font = UIFont.systemFont(ofSize: textSize)
        let paragraph = NSMutableParagraphStyle()
        paragraph.alignment = .center

        for (index, char) in text.enumerated() {
            let centerX = startX + CGFloat(index) * characterSpace
            let circleColor = characterBackgroundColors?[safe: index] ?? characterBackgroundColor

            // Draw circle
            let circlePath = UIBezierPath(arcCenter: CGPoint(x: centerX, y: centerY),
                                          radius: radius,
                                          startAngle: 0,
                                          endAngle: CGFloat.pi * 2,
                                          clockwise: true)
            circleColor.setFill()
            circlePath.fill()

            // Draw text
            let attributes: [NSAttributedString.Key: Any] = [
                .font: font,
                .foregroundColor: textColor,
                .paragraphStyle: paragraph
            ]

            let charString = String(char)
            let size = charString.size(withAttributes: attributes)
            let textY = centerY - size.height / 2
            charString.draw(at: CGPoint(x: centerX - size.width / 2, y: textY), withAttributes: attributes)
        }
    }
}
